File: c/yacht/yacht.c
#include "yacht.h"

static void count_dice(dice_t dice, int count[NUM_FACES]) {
    for (int i = 0; i < NUM_FACES; i++) {
        count[i] = 0;
    }

    for (int i = 0; i < NUM_DICE; i++) {
        count[dice.faces[i] - 1]++;
    }
}

static int score_num(dice_t dice, int num) {
    int ans = 0;
    for (int i = 0; i < NUM_DICE; i++) {
        if (num == -1) ans += dice.faces[i];
        if (dice.faces[i] == num) ans += num;
    }
    return ans;
}

static int score_full_house(dice_t dice) {
    int count[NUM_FACES];
    count_dice(dice, count);

    int ans = 0;
    for (int i = 0; i < NUM_FACES; i++) {
        if (count[i] == 1 || count[i] == 4 || count[i] == 5) return 0;
        ans += (i + 1) * count[i];
    }
    return ans;
}

static int score_four_of_a_kind(dice_t dice) {
    int count[NUM_FACES];
    count_dice(dice, count);

    for (int i = 0; i < NUM_FACES; i++) {
        if (count[i] == 4 || count[i] == 5) return (i + 1) * 4;
    }
    return 0;
}

static int score_yatch(dice_t dice) {
    int count[NUM_FACES];
    count_dice(dice, count);

    for (int i = 0; i < NUM_FACES; i++) {
        if (count[i] == 5) return 50;
    }
    return 0;
}

static int score_straight(dice_t dice, char big) {
    int count[NUM_FACES];
    count_dice(dice, count);

    for (int i = 0; i < NUM_FACES; i++) {
        if (count[i] != 1 && count[i] != 0) return 0;
    }
    if ((big == 0 && count[NUM_FACES - 1] == 0) ||
        (big != 0 && count[0] == 0)) return 30;
    return 0;
}

int score(dice_t dice, category_t category) {
    switch (category) {
        case ONES:
        case TWOS:
        case THREES:
        case FOURS:
        case FIVES:
        case SIXES:
            return score_num(dice, category + 1);
        case FULL_HOUSE:
            return score_full_house(dice);
        case FOUR_OF_A_KIND:
            return score_four_of_a_kind(dice);
        case LITTLE_STRAIGHT:
        case BIG_STRAIGHT:
            return score_straight(dice, category - LITTLE_STRAIGHT);
        case CHOICE:
            return score_num(dice, -1);
        case YACHT:
            return score_yatch(dice);
        default:
            return 0;
    }
    return 0;
}

File: zig/yacht/yacht.zig
pub const Category = enum {
    ones,
    twos,
    threes,
    fours,
    fives,
    sixes,
    full_house,
    four_of_a_kind,
    little_straight,
    big_straight,
    choice,
    yacht,
};

fn count_dice(dice: [5]u3) [6]u3 {
    var count: [6]u3 = .{ 0, 0, 0, 0, 0, 0 };
    for (dice) |d| {
        count[d - 1] += 1;
    }
    return count;
}

fn score_num(dice: [5]u3, num: i5) u32 {
    var ans: u32 = 0;
    for (dice) |d| {
        if (num == -1) ans += d;
        if (d == num) ans += @intCast(num);
    }
    return ans;
}

fn score_full_house(dice: [5]u3) u32 {
    const count = count_dice(dice);

    var ans: u32 = 0;
    for (0..6) |i| {
        if (count[i] == 1 or count[i] == 4 or count[i] == 5) return 0;
        ans += @truncate((i + 1) * count[i]);
    }
    return ans;
}

fn score_four_of_a_kind(dice: [5]u3) u32 {
    const count = count_dice(dice);

    for (0..6) |i| {
        if (count[i] == 4 or count[i] == 5) return @truncate((i + 1) * 4);
    }
    return 0;
}

fn score_yatch(dice: [5]u3) u32 {
    const count = count_dice(dice);

    for (0..6) |i| {
        if (count[i] == 5) return 50;
    }
    return 0;
}

fn score_straight(dice: [5]u3, category: Category) u32 {
    const count = count_dice(dice);

    for (0..6) |i| {
        if (count[i] != 1 and count[i] != 0) return 0;
    }
    if ((category == .little_straight and count[5] == 0) or
        (category == .big_straight and count[0] == 0)) return 30;
    return 0;
}

pub fn score(dice: [5]u3, category: Category) u32 {
    return switch (category) {
        .ones, .twos, .threes, .fours, .fives, .sixes => score_num(dice, @intFromEnum(category) + 1),
        .full_house => score_full_house(dice),
        .four_of_a_kind => score_four_of_a_kind(dice),
        .little_straight, .big_straight => score_straight(dice, category),
        .choice => score_num(dice, -1),
        .yacht => score_yatch(dice),
    };
}
